File: DSA/Aarush_590017879/Day_12/Q2.c
#include <stdio.h>
#include <string.h>
#include <limits.h>

int smallestDistinctWindow(char *s) {
    int n = strlen(s);

    // Count total distinct characters
    int visited[256] = {0};
    int distinct = 0;

    for (int i = 0; i < n; i++) {
        if (!visited[(unsigned char)s[i]]) {
            visited[(unsigned char)s[i]] = 1;
            distinct++;
        }
    }

    int freq[256] = {0};
    int left = 0, count = 0;
    int minLen = INT_MAX;

    for (int right = 0; right < n; right++) {
        if (freq[(unsigned char)s[right]] == 0)
            count++;

        freq[(unsigned char)s[right]]++;

        while (count == distinct) {
            if (right - left + 1 < minLen)
                minLen = right - left + 1;

            freq[(unsigned char)s[left]]--;

            if (freq[(unsigned char)s[left]] == 0)
                count--;

            left++;
        }
    }

    return minLen;
}

int main() {
    char s[1001];

    scanf("%s", s);

    printf("%d\n", smallestDistinctWindow(s));

    return 0;
}

File: DSA/vaishnavi_joshi_590016947/Day12/Q2.java
import java.util.*;

public class SmallDistinctWindow {

    public static int findSubString(String str) {
        int n = str.length();

        // Count total distinct characters
        boolean[] visited = new boolean[256];
        int distinct = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[str.charAt(i)]) {
                visited[str.charAt(i)] = true;
                distinct++;
            }
        }

        int[] freq = new int[256];
        int start = 0;
        int count = 0;
        int minLen = Integer.MAX_VALUE;

        for (int end = 0; end < n; end++) {
            char ch = str.charAt(end);

            freq[ch]++;

            if (freq[ch] == 1)
                count++;

            while (count == distinct) {
                minLen = Math.min(minLen, end - start + 1);

                freq[str.charAt(start)]--;

                if (freq[str.charAt(start)] == 0)
                    count--;

                start++;
            }
        }

        return minLen;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Input String
        String str = sc.nextLine();

        System.out.println(findSubString(str));

        sc.close();
    }
}